osu.Game/Screens/OnlinePlay/Components/RoomManager.cs

```diff
@@ -87,7 +87,8 @@ public virtual void JoinRoom(Room room, string password = null, Action<Room> onS
 
             currentJoinRoomRequest.Failure += exception =>
             {
-                if (!(exception is OperationCanceledException))
+                // provide error output if the operation wasn't canceled and the error doesn't stem from an invalid password
+                if (!(exception is OperationCanceledException) && !((APIException)exception).Message.Equals("Invalid room password", StringComparison.Ordinal))
```

Comment:
This won't work if the user's language on the website is not in English. [The API error message is localised.](https://github.com/ppy/osu-web/blob/ee33f9215579f397824b7ebf074e0f0b7d27a3da/app/Http/Controllers/Multiplayer/RoomsController.php#L56)

This should either check the response error code, or web should return a language-invariant error "type" token of some kind.

Reply:
[The function that handles authentication for rooms](https://github.com/ppy/osu-web/blob/ee33f9215579f397824b7ebf074e0f0b7d27a3da/app/Http/Controllers/Multiplayer/RoomsController.php#L43) that you linked to returns `403` for different reasons (if the user being added to the room is not the currently authenticated user & invalid password). I may have misunderstood what the [first `403` returns](https://github.com/ppy/osu-web/blob/ee33f9215579f397824b7ebf074e0f0b7d27a3da/app/Http/Controllers/Multiplayer/RoomsController.php#L47) as I'm not too familiar with PHP, but judging from the naming that looks like what the check is.

Because of the varying `403` responses, having the feedback be the same for an unauthenticated user/invalid user id and an invalid password may be confusing.

A `type` token with the response would work, or a `401 Unauthorized` could be returned instead of a `403` for an invalid password. Either solution works, but would require updates to web before this PR can be merged.

Reply:
401 is wrong. It means "no credentials provided", and 403 means "credentials provided are wrong".

Also, when I attempted to add test coverage of this, the hard cast to `APIException` will cause crashes if `exception` is not indeed of that type. This should be using a soft cast like the `OperationCanceledException` case does.

Reply:
> 401 is wrong. It means "no credentials provided", and 403 means "credentials provided are wrong".

Misread the docs for `401`, my bad!

> Also, when I attempted to add test coverage of this, the hard cast to APIException will cause crashes if exception is not indeed of that type. This should be using a soft cast like the OperationCanceledException case does.

Makes sense, the `if` should be checking if `exception` is actually an `APIException` before trying to cast to it.

---

Also, assuming I'm right about the [first `403` return](https://github.com/ppy/osu-web/blob/ee33f9215579f397824b7ebf074e0f0b7d27a3da/app/Http/Controllers/Multiplayer/RoomsController.php#L47), would it actually just be safe to check against a `403` and assume that the `userId` provided by lazer would be the same as the authenticated user? My main hesitation is that I don't want the user to receive the same feedback if their password is wrong (user's fault) vs. lazer is sending the wrong `userId` in the request (internal issue).

Reply:
Getting error information from the source of the error (i.e. the API/web) is always preferred to implementing heuristics in order to discern what an error means client-side. There's no reason to make changes to web to be able to tell what the error means in a user-language-agnostic manner.

Reply:
I'm not entirely sure how the `abort()` method works in osu-web, so I'm unsure on how to go about returning a `type` token in the body. If possible, the `DisplayableError` class could have a nullable `JsonProperty` containing the `type` token (as I assume no other endpoint would return this token), which can be checked instead of the message.

Reply:
This implementation also just fails when trying to add a test (which this PR is missing). @sh0ckR6 we require that tests are added to cover new behaviours like this for this precise reason.